Port sort key code from ICU4C (#2689)

rs-sac opened the pull request:
I mostly translate the C++ code straight across, but I made some minor changes to match the mindset of the Rust codebase, to match Rust practice, to satisfy Clippy, and to reduce copy-and-paste.

A comment on the issue suggested handling the identical level by appending UTF-8.  That's not what ICU4C does, but it's a whole lot simpler, so that's what I did.

At the end, some simple tests are included.

Fixes #2689

rs-sac commented:
I see that most of the test failures boil down to my calling `new_nfd` outside of a `#[cfg(feature = "compiled_data")]`.  If you could advise me on what I did wrong (call something else, add the feature attribute there, do something else), I'll endeavor to fix it.

Manishearth commented:
(I don't tend to review collator stuff)

hsivonen commented:
Thank you for the PR!

I acknowledge that this is my area to review, but I couldn't get to it today.

sffc commented:
Thanks for the contribution! A few things:

1. I received confirmation that your Corporate CLA was received. You can select the appropriate option for yourself on cla-assistant.io.
2. This PR was opened just hours before #6496 landed, which will need to be merged/rebased.
3. About the data provider CI errors: if you call any function that is `#[cfg(feature = "compiled_data")]`, then your function also needs to be `#[cfg(feature = "compiled_data")]`, and it should have a corresponding `_with_buffer_provider` and `_unstable`. We usually have an internal function that takes the resolved objects and then we write these three functions as thin wrappers on top of the internal function.

rs-sac commented:
Thanks for the pointers.  I think I have fixed the problem, though I await output from the official tests.

rs-sac commented:
Rather than add to the temporary diplomat exclusion file, I added my stuff to the more documented allow list.

I'm not an expert on diplomat, but I'm pretty sure it doesn't make sense to expose the internal use constructor I created.  OTOH, ideally the sort key code would be exposed, but AFAICT as a non-expert, that can't happen due to using traits.

(Since I don't know what reviewers are necessary, I didn't add or change any reviewers, even though this web site says I did.  It's completely making that up.)

hsivonen commented:
> However, I think the data you need might already be there. `CollatorBorrowed` already contains a field `decompositions: &'a DecompositionData<'a>`, which you can use to create a `Decomposition` (see line 594) and presumably also a `DecomposingNormalizerBorrowed`.

The data is indeed already there. [The collator already creates `Decomposition` using the data it already has.](https://github.com/unicode-org/icu4x/blob/17a23af302b30cfe4a9f19d870f28888e6566d35/components/collator/src/comparison.rs#L533)

It looks like a `pub` (but probably doc-hidden) constructor for creating a `DecomposingNormalizerBorrowed` from the data that the collator already holds is needed (plugging `None` and the correct `u8` constants for NFD in the fields that need input not held by the collator).

hsivonen commented on components/collator/src/comparison.rs:
```diff
@@ -772,36 +812,97 @@ impl CollatorBorrowed<'_> {
+        }
+    }
 
+    fn collation_elements<'a, I>(
```
This should optimize to not causing a move of `CollationElements`, but rerunning benches is needed to confirm.

rs-sac replied:
This was a regression.  Inlining wasn't enough to fix it; I had to macroize it.

hsivonen commented:
> > However, I think the data you need might already be there. `CollatorBorrowed` already contains a field `decompositions: &'a DecompositionData<'a>`, which you can use to create a `Decomposition` (see line 594) and presumably also a `DecomposingNormalizerBorrowed`.
> 
> The data is indeed already there. [The collator already creates `Decomposition` using the data it already has.](https://github.com/unicode-org/icu4x/blob/17a23af302b30cfe4a9f19d870f28888e6566d35/components/collator/src/comparison.rs#L533)
> 
> It looks like a `pub` (but probably doc-hidden) constructor for creating a `DecomposingNormalizerBorrowed` from the data that the collator already holds is needed (plugging `None` and the correct `u8` constants for NFD in the fields that need input not held by the collator).

Ah, nevermind, it's already in the changeset.

hsivonen commented on components/collator/src/comparison.rs:
```diff
@@ -1543,4 +1644,660 @@ impl CollatorBorrowed<'_> {
+                DecomposingNormalizerBorrowed::new_with_data(self.decompositions, self.tables);
+            let s = nfd.normalize(s);
+            sink.write_byte(LEVEL_SEPARATOR_BYTE);
+            sink.write(s.as_bytes());
```
It would be better not to allocate the intermediate string but to use `nfd.normalize_to(s, sink)` with the `Write` trait declared in this changeset extending `core::fmt::Write` so that `sink` is suitable for `normalize_to`.

rs-sac replied:
You are right about the temporary, but how is this possible?  `Vec<u8>` is `std::io::Write` but not `core::fmt::Write`.  Do you envision that the concrete type here would be some type other than Vec/SmallVec?

hsivonen replied:
I was thinking of the concrete type being `Vec<u8>`, but I guess the rules don't allow what I had in mind.

How about this:

```rust
struct SinkAdapter<'a, W: crate::comparison::Write> {
    delegate: &'a mut W,
}

impl<'a, W: crate::comparison::Write> SinkAdapter<'a, W> {
    fn new(delegate: &'a mut W) -> Self {
        Self {
            delegate,
        }
    }
}

impl<'a, W: crate::comparison::Write> core::fmt::Write for SinkAdapter<'a, W> {
    fn write_str(&mut self, s: &str) -> core::fmt::Result {
        self.delegate.write(s.as_bytes());
        Ok(())
    }
}
```

and then

```rust
            let nfd =
                DecomposingNormalizerBorrowed::new_with_data(self.decompositions, self.tables);
            sink.write_byte(LEVEL_SEPARATOR_BYTE);
            let mut adapter = SinkAdapter::new(sink);
            let _ = nfd.normalize_to(s, &mut adapter);
```

rs-sac replied:
Done.

hsivonen commented on components/normalizer/src/lib.rs:
```diff
@@ -1755,6 +1755,24 @@ impl DecomposingNormalizerBorrowed<'static> {
+    /// This constructor is intended for use by collations.
+    ///
+    /// [📚 Help choosing a constructor](icu_provider::constructors)
+    pub fn new_with_data(
```
Let's make this `#[doc(hidden)]`.

rs-sac replied:
Done.

hsivonen commented on components/collator/src/comparison.rs:
```diff
@@ -1543,4 +1644,660 @@ impl CollatorBorrowed<'_> {
+    /// Optionally write the case level.  Stop writing levels when `need_to_write` returns
+    /// `false`.  Separate levels with the `LEVEL_SEPARATOR_BYTE`, but do not write a
+    /// `TERMINATOR_BYTE`.
+    pub fn write_sort_key_up_to_quaternary<I, S, W>(&self, iter: I, sink: &mut S, need_to_write: W)
```
It's a bit unclear to me how and why an application would use the `need_to_write` callback, so I'm not sure I can assess right now whether this method should be private for now.

AFAICT, the callback isn't part of ICU4C's public API. Am I missing something?

rs-sac replied:
Removed.

hsivonen commented on components/collator/src/comparison.rs:
```diff
@@ -1543,4 +1644,660 @@ impl CollatorBorrowed<'_> {
+                    prev_reordered_primary = if is_compressible { p } else { 0 };
+                }
+
+                let p2 = (p >> 16) as u16;
```
Shouldn't this narrow to `u8` instead of narrowing to `u16`?

rs-sac replied:
Yes, good catch.

hsivonen commented on components/collator/src/comparison.rs:
```diff
@@ -1543,4 +1644,660 @@ impl CollatorBorrowed<'_> {
+        self.write(&[b]);
+    }
+
+    /// Write leading bytes up to a zero byte, but always write at least one byte.
```
AFAICT, this is always called with an array of three bytes whose first byte is known to be non-zero, so do we need to check here for writing at least one byte?

Also, wouldn't it be better to have this as a helper function outside the trait so that trait implementors don't get to override this method with a broken implementation?

hsivonen replied:
Since there's one call site with at most 3 bytes and it already checks the first byte for zero, it seems to me that nested `if`s to check the other two bytes and writing each using a per-byte call would work.

hsivonen commented on components/collator/src/comparison.rs:
```diff
@@ -1543,4 +1644,660 @@ impl CollatorBorrowed<'_> {
+                } else {
+                    // case bits & tertiary lead byte
+                    let mut c = ((non_primary.bits() >> 8) & 0xff) as u8;
+                    assert_ne!(c & 0xc0, 0xc0);
```
Please downgrade the assertions to debug assertions. It seems to me that the assertions added in this PR are not needed for memory-safety, and we generally want to avoid panics even if the alternative is bogus output.

hsivonen commented on components/collator/src/comparison.rs:
```diff
@@ -1543,4 +1644,660 @@ impl CollatorBorrowed<'_> {
+                            // mixed=14, upper=15.  If there are only common (=lowest) weights
+                            // in the whole level, then we need not write anything.  Level
+                            // length differences are handled already on the next-higher level.
+                            if common_cases != 0 && (c > LEVEL_SEPARATOR_BYTE || cases.is_empty()) {
```
This should have a negation `!` before `cases.is_empty()`, right?

rs-sac replied:
Yes, good catch.

hsivonen commented on components/collator/src/comparison.rs:
```diff
@@ -1543,4 +1644,660 @@ impl CollatorBorrowed<'_> {
+            sink.write(s.as_bytes());
+        }
+
+        sink.write(&[TERMINATOR_BYTE]);
```
This is just for C string compatibility, right?

I'm inclined to think that we shouldn't append the zero terminator for callers that use the output as a buffer that knows its length, but we should document that if the input doesn't contain U+0000, it's OK for the caller to append `0u8` to the sink after this method returns and then treat the result as a sentinel-based C string.

rs-sac replied:
Removed.

hsivonen commented on components/collator/src/comparison.rs:
```diff
@@ -1543,4 +1644,660 @@ impl CollatorBorrowed<'_> {
+    }
+
+    /// Write the sort key bytes up to the collator's strength.
+    ///
```
Before release, this needs documentation covering at least:

* A non-key comparison with the collator typically entails a single primary-level comparison after code unit-level identical prefix skipping, so it takes a lot of comparisons to amortize the performance cost of generating collation keys. Applications should not use collation keys as a performance optimization without measuring that the performance cost of generating the collation keys is actually amortized.
* The stored collation keys should be presumed to become invalidated by a CLDR update, a Unicode Database update (i.e. new version of Unicode), or an update to the ICU4X code. Applications must, therefore, be prepared to recompute all collation keys and take the performance cost of doing so into account when considering whether the performance cost of collation keys is amortized.
* It has happened more than once in the past with other collator implementations that a database has stored collation keys and then the database has grown so large that regenerating the collation keys has become too expensive, which has then resulted in engineering effort to use a frozen copy of collation key generation algorithm and the associated data while making other updates to software. Users of ICU4X should not expect ICU4X to accommodate this kind of need if a user of ICU4X gets into this situation despite this warning.

Adding the warnings can be a follow-up, though.

rs-sac replied:
Done.

hsivonen reviewed (changes requested):
Generally looks great, but I have some change requests inline.

I think the biggest concern is planning correctly around the compressibility table that ICU4X currently doesn't have.

I intend to discuss that aspect with other ICU4X developers on Thursday.

hsivonen commented on components/collator/src/comparison.rs:
```diff
@@ -1543,4 +1644,660 @@ impl CollatorBorrowed<'_> {
+            // primary level but terminate compression on all levels and then exit the loop.
+            if p > NO_CE_PRIMARY && levels & PRIMARY_LEVEL_FLAG != 0 {
+                // Test the un-reordered primary for compressibility.
+                let is_compressible = false; // TODO?
```
I've thought about this more. Here's what I think we should (not necessarily in this PR but before releasing without a non-experimental flag):

* Make icuexportdata export 256 bits of `compressibleBytes` data for the root. (AFAICT, only the root has these bits.) Perhaps export this as an array of 256 bytes each of which is 1 or 0 without packing into an array of four 64-bit integers on the ICU4C side and leave the packing to datagen.
* Introduce `CollationKeyGeneratorBorrowed` and `CollationKeyGenerator` that hold a `CollatorBorrowed` plus `& [u64; 4]` and `Collator` plus `[u64; 4]`, respectively.
* Introduce a singleton data key for compressibly bytes, which is a table of 256 bits represented as `[u64; 4]`. 
* Put the public entry points for collation key generation on `CollationKeyGeneratorBorrowed`.

hsivonen replied:
> * and `Collator` plus `[u64; 4]`,

With the [u64; 4] inside a data struct.

rs-sac replied:
The data are added and integrated.

rs-sac commented:
> 1. Should the sink trait be called `Write`, because traits roughly like that are customarily called `Write`, or should the trait be called something else so that it's obvious from the unqualified name that it's not one of the other `Write`s?

As a newbie, I don't have a strong opinion.  I took the name suggested in the original issue.

> 2. Should the methods on the trait return `Result` even though when implemented for `Vec` and `SmallVec`, the implementation would always return `Ok(())`? Is it valuable to be able to write fallible sinks when the internal `SmallVec`s can panic on OOM anyway?

I had the same thoughts.  It appeared to me that while the C++ interface supports errors, they were all related to OOM, which is just a crash in Rust (lame IMO, but nobody asked my opinion on that point).  Since I didn't see any plausible scenario wherein an error would be returned, I skipped the error business, thinking that if I'd overlooked anything, it would be pointed out in review.

Thanks for the rest of the comments.  It will take me a little while to work through them all.

hsivonen commented:
From the meeting today:

- @hsivonen Overall, high quality. I found two cases of what appears to be minor typos. There are three things I'd like input on. The most substantive one is, this will need 256 bits of data exported from ICU4C into a new kind of data struct that we don't currently have. If we follow the ICU4X philosophy properly, then it would mean that this method would need to be on a different object, implying things like `CollationKeyGenerator`, `CollationKeyGeneratorBorrowed`. What should we do in this case?
- @sffc 256 bits, not even bytes?
- @hsivonen Yes. ICU4C uses bytes, but we can use bits.
- @robertbastian I think we should add a new data marker, update the unstable constructor, and fall back to hardcoded data in the buffer constructor
- @hsivonen We should ask @markusicu to verify whether the algorithm still works if we say if `is_compressible` is always false. And we should also verify what happens if you generate collation keys with the compressible data and without the compressible data. I'm not sure how the compression works, so I can't say. I think we should have a singleton data key for the compressibility of the bytes, and that's a data struct that contains an array of four u64s. If we fail to load that data struct, do we fail the construction, or do we will in all zeros? If we really want ICU4X 2.x to ingest 2.0 data that doesn't have this, and construct successfully, we have to say to fill in all zeros. But I think it's better to say that if the data key doesn't resolve, you don't get to produce collation keys, which is probably the better way forward.
- @sffc Ideally we would just include these 256 bits in the main data struct. We don't have a black and white rule on when we split things into different data structs and different types, but generally if something needs data that is very small (less than ~1 KB, certainly less than ~100 B) and closely related to the functionality of the main type, it can just go onto the main type.
- @hsivonen We could make a `SpecialPrimariesV2` with the extra field.
- @robertbastian we're allowed to add a new marker, and remove the old marker from the unstable constructor. our plan for custom baked data and buffer data is to keep the old key around, but given that we only released a week ago, it's *probably* fine to not do that
- @hsivonen What zerovec type should I use? Do we have a fixed-length array?
- @sffc You can use a `VarZeroCow<(...)>` if you want
- @robertbastian I'd just use a `ZeroVec<u8>`
- @hsivonen I'll use `ZeroVec<u8>`
- @sffc Did we decide what to actually do?
- @hsivonen I'll add a field `compressible_bytes` in `SpecialPrimaries` struct, and I update the marker name from `V1` to `V2`.
- @sffc Yeah, we should be strict with our semver policy, and the exception we're making is only in our "new code old data" policy, which I'm comfortable making.
- @robertbastian intialize `compressible_bytes` with the CLDR 47 data in datagen if icuexportdata doesn't contain the data. then in the future add it to icuexportdata

hsivonen commented on components/collator/src/comparison.rs:
```diff
@@ -1543,4 +1644,660 @@ impl CollatorBorrowed<'_> {
+        // // License & terms of use: http://www.unicode.org/copyright.html
+        // /*
+        // *******************************************************************************
+        // * Copyright (C) 2012-2015, International Business Machines
```
@srl295 Given that we have the IBM copyright notice with a broader year range already in https://github.com/unicode-org/icu4x/blob/09d75a966aaeb2a302f195ce4a91bfa8a8eb5d1b/LICENSE#L45 , do we need this license block to state the license header of the relevant ICU4C file, or can this whole block comment be shortened to something like:

"The algorithm comes from `CollationKeys::writeSortKeyUpToQuaternary` in ICU4C."

hsivonen commented:
> * @sffc Maybe use `core::fmt::Error` as the error type instead of creating a new one?

It occurred to me after the meeting that `SinkAdapter` that I proposed upthread wouldn't need a conversion between error types if the error type is the same as in the normalizer, so that's in favor of using `core::fmt::Error`.

sffc commented:
Thought about semver: there's a potential issue if the resolver bumps `icu_collator` to 2.0.1 but not with the corresponding change in `icu_collator_data` (old-code-new-data) and vice-versa. This is why we have `~` deps.

Is it a problem here? Probably not. But, we have some paths to mitigate available to us:

1. We should definitely make sure `icu_collator 2.0.1` depends on `icu_collator_data ~2.0.1` as opposed to `~2.0`. This prevents new-code-old-data problems with compiled_data.
2. If someone runs cargo update and they somehow get broken with old-code-new-data, we can always release `icu_collator_data 2.0.2` that contains both markers.

rs-sac commented on components/collator/src/comparison.rs:
```diff
@@ -1543,4 +1644,660 @@ impl CollatorBorrowed<'_> {
+    ///
+    /// For identical strength, the UTF-8 NFD normalization is appended for breaking ties.
+    /// The key ends with a `TERMINATOR_BYTE`.
+    pub fn write_sort_key<S>(&self, s: &str, sink: &mut S)
```
Done.

rs-sac commented:
Due to my putting a CHANGELOG entry into my patches, I had to rebase.  I removed the entry because it sounded as if you folks would probably prefer to word it yourselves.  The rest of the patches are the same.  Then I added more commits on top because I read that's the style you prefer here.

I believe I have addressed all of the feedback, including the comments from the meeting about the trait's name and fallibility.  I left your comments open in case you wanted to only resolve them after you check the new patches.

Regarding upgrades, you might want to think about how to communicate potential sort key incompatibility to users.  Should that be via semver?  I added the requested dire warnings about regeneration, but it seems a little unreasonable if there's no guarantee at all about compatibility between versions X.Y.0 and X.Y.1.

hsivonen commented:
#6604 adds the compressibility data, hopefully correctly.

hsivonen commented on components/collator/src/comparison.rs:
```diff
@@ -1543,4 +1631,808 @@ impl CollatorBorrowed<'_> {
+        S: CollationKeySink,
+    {
+        let nfd = DecomposingNormalizerBorrowed::new_with_data(self.decompositions, self.tables);
+        let iter = nfd.normalize_iter(s.chars());
```
Likewise, this iterator wrapping is unnecessary.

hsivonen commented on components/collator/src/comparison.rs:
```diff
@@ -1543,4 +1631,808 @@ impl CollatorBorrowed<'_> {
+
+    // guaranteed to get at least one byte and the rest are checked incrementally
+    #[allow(clippy::indexing_slicing)]
+    fn write_to_zero(&mut self, buf: &[u8]) -> Result<(), core::fmt::Error> {
```
Please inline this to the single (non-test) call site, so that this can be removed from the trait.

hsivonen commented on components/collator/src/comparison.rs:
```diff
@@ -1543,4 +1631,808 @@ impl CollatorBorrowed<'_> {
+        for c in iter {
+            let c = c.unwrap_or(char::REPLACEMENT_CHARACTER); // shouldn't happen
+            let len = c.len_utf8();
+            let mut bytes = vec![0u8; len];
```
Instead of allocating a `Vec`, please use an on-stack `[0u8; 4]` as the buffer for `encode_utf8`.

hsivonen commented on components/collator/src/comparison.rs:
```diff
@@ -1543,4 +1631,808 @@ impl CollatorBorrowed<'_> {
+impl<W> write16::Write16 for SinkAdapter<'_, W>
+where
+    W: CollationKeySink,
+{
```
As a micro optimization, please provide an implementation like this to encode single chars directly to UTF-8:

```rust
    fn write_char(&mut self, c: char) -> core::fmt::Result {
        let mut buf = [0u8; 4];
        self.inner.write(c.encode_utf8(&mut buf).as_bytes())
    }
```

hsivonen commented on components/collator/src/comparison.rs:
```diff
@@ -1543,4 +1631,781 @@ impl CollatorBorrowed<'_> {
+        let iter = char::decode_utf16(s.iter().cloned());
+        for c in iter {
+            let c = c.unwrap_or(char::REPLACEMENT_CHARACTER); // shouldn't happen
+            let mut bytes = [0u8; 4];
```
Hoisting this out of the loop manually might be a micro optimization, since the exact zeroing semantics might prevent LLVM from hoisting it. Anyway, not an essential thing.

hsivonen commented:
I'll be unavailable tomorrow and for most of next week, so I'm marking this approved in order to unblock landing once #6604 has landed. There are two things remaining:

1. Adding the query to `compressible_bytes`
2. @srl295 's take on the specifics of marking the provenance as port from ICU4C

rs-sac commented:
• Rebase to pick up compressible bytes patch
• Adjusted my refactoring patch for rebase
• Added the is_compressible call

@srl295, do you have specific requirements regarding crediting ICU4C as the source of the algorithm?

sffc commented on components/collator/src/comparison.rs:
```diff
@@ -1551,4 +1633,771 @@ impl CollatorBorrowed<'_> {
+/// (This crate does not have access to [`std`].)
+pub trait CollationKeySink {
+    /// Writes a buffer into the writer.
+    fn write(&mut self, buf: &[u8]) -> Result<(), core::fmt::Error>;
```
I prefer `core::fmt::Error`. My thinking is:

1. This trait is intended to closely model the `Write` traits in core/std
2. Both implementation code and client code is already built to propagate `core::fmt::Error`
3. Let's not export extra types when we don't need to; we are exporting `CollationKeySink` because we need to, but that doesn't mean we should export its own little write-error type

And I think it should be fallible:

1. It's easy to imagine use cases with fixed-length buffers. Maybe you are trying to save your collation key into a fixed-length field in a database.
2. And of course because we are trying to model core/std

sffc commented on tools/make/diplomat-coverage/src/allowlist.rs:
```diff
@@ -221,6 +224,11 @@ lazy_static::lazy_static! {
         "icu::normalizer::Composition",
         "icu::normalizer::Decomposition",
 
+        // Trait required
+        "icu::collator::CollatorBorrowed::write_sort_key",
+        "icu::collator::CollatorBorrowed::write_sort_key_utf8",
+        "icu::collator::CollatorBorrowed::write_sort_key_utf16",
```
+1

robertbastian replied:
There's definitely a way to make this work, details can be hashed out later.

For this PR I'd like to remove them from the allowlist, as they are actually missing APIs that should be listed in `missing_apis.txt`.

rs-sac commented on components/collator/src/comparison.rs:
```diff
@@ -1551,4 +1633,771 @@ impl CollatorBorrowed<'_> {
+    fn write_slice(&mut self, s: &[u16]) -> core::fmt::Result {
+        // For the identical level, if the input is UTF-16, transcode to UTF-8.
+        let iter = char::decode_utf16(s.iter().cloned());
+        let mut bytes = [0u8; 4];
+        for c in iter {
+            let c = c.unwrap_or(char::REPLACEMENT_CHARACTER); // shouldn't happen
+            self.inner.write(c.encode_utf8(&mut bytes).as_bytes())?;
+        }
```
Would that reinitialize the array every time?

rs-sac commented on components/collator/src/comparison.rs:
```diff
@@ -1551,4 +1633,771 @@ impl CollatorBorrowed<'_> {
+                    if 0 < p && p <= MERGE_SEPARATOR_PRIMARY {
+                        // The backwards secondary level compares secondary weights backwards
+                        // within segments separated by the merge separator (U+FFFE).
+                        let secs = secondaries.as_mut();
```
Done.

robertbastian commented on components/collator/src/comparison.rs:
```diff
@@ -1551,4 +1633,771 @@ impl CollatorBorrowed<'_> {
+    }
+
+    fn write_char(&mut self, c: char) -> core::fmt::Result {
+        let mut buf = [0u8; 4];
+        self.inner.write(c.encode_utf8(&mut buf).as_bytes())
```
this isn't in a loop

rs-sac replied:
That's true.  Done.

robertbastian commented on components/normalizer/src/lib.rs:
```diff
@@ -1755,6 +1755,25 @@ impl DecomposingNormalizerBorrowed<'static> {
+    ///
+    /// [📚 Help choosing a constructor](icu_provider::constructors)
+    #[doc(hidden)]
+    pub fn new_with_data(
```
Ah I see. We should clean this up (Collator should own a DecomposingNormalizer instead of the parts of it), but that can be in a different PR

Manishearth commented:
Discussion from today's ICU4X WG meeting

<details><summary> Meeting notes </summary>


- @sffc (introduces the issue)
- @robertbastian There are two problems. First, should we use a sink or return an iterator? Second, if we use a sink, what is the error type?
- @Manishearth What are the error cases?
- @robertbastian They are sink errors.
- @robertbastian I think the reason `core::fmt::Write` exists is because returning an iterator of chars is an overhead. Returning an iterator of bytes should not be an overhead compared to writing to a byte sink.
- @Manishearth I think it isn't about overhead more than API choice. But you're right that there's no difference, but also `std::io::Write` exists. So it's more about, what do peope want here? Here's a question. We're treating this as a collation sink. A property it has is that the Write trait can append onto something existing. Is that something we care about here?
- @robertbastian I think sort keys are homomorphic, right?
- @Manishearth So you might end up writing to an existing string? Because if not, we could return FromIterator Bytes.
- @sffc Normally we return Writeable types. Should we do that here?
- @robertbastian In Normalizer we use the sink model.
- @robertbastian Can you clarify how we could return a FromIterator?
- @Manishearth Instead of returning an Iterator over bytes, you return a T where T: FromIterator Bytes, and you specify the T based on context.
- @robertbastian That seems bad because then you can't use extend
- @Manishearth Yeah. I was feeling out the design space.
- @Manishearth We return Writeable for the purpose of, you need to do a bunch of calculation. Separating out the writing task from the collection task. I'm not sure that's a problem here. I don't think we have much of an API benefit here.
- @robertbastian - we don't do this two-step thing with intermediate writeable types in normalizer: https://docs.rs/icu/latest/icu/normalizer/struct.ComposingNormalizerBorrowed.html#method.normalize_to
- @robertbastian I don't like duplicating a trait just because we're no_std, and besides, `std::io::Write` isn't the right trait because `std::io::Error` is not the right type.
- @Manishearth I think it's barely the right trait.
- @sffc Is there a reason why in Normalizer we take a sink instead of returning a Writeable?
- @robertbastian I didn't notice this until this week.
- @Manishearth There's another wrinkle, which is that for `&mut`, you need to return filled bytes. I think the resolution is that the fn can return a `usize` of the number of bytes written.
- @robertbastian - at that point I'd prefer a second method call for this
- @sffc Would it help if we added `writeable::WriteableBytes`?
- @Manishearth `Writeable` is a replacement for `Display`. Something that looks like `Writeable` doesn't help us here. If we had `WriteableBytes`, we'd need `WriteBytes`.
- @sffc Another thing we could do is return a `WriteableSortKey` that doesn't implement `Writeable` but has its own associated functions that are similar in shape, and implements `IntoIterator<u8>`
- @robertbastian What would the associated functions be?
- @sffc It would be `into_iter()` and `write_to()` which still takes the sink trait, but then users can pick which one works better
- @Manishearth I think this is solving a different problem. `WriteableBytes` is orthogonal. `into_iter()` and `write_to()` are equally powerful APIs; there is no need to have both. The benefit of Writeable doesn't really translate to this world.
- @robertbastian The problem with adding a sink trait is that we need to provide all the impls for standard library types, which is annoying. Also, we cannot implement it for 3P types, so callers would often need newtypes and write wrapper impls
- @sffc If we returned an iterator, it looks like `extend` can be used to append to a vec: https://doc.rust-lang.org/std/vec/struct.Vec.html#impl-Extend%3C%26T%3E-for-Vec%3CT,+A%3E
- @Manishearth I think they're moving `std::io::Write` to `core`
- @robertbastian It's still aways off

Currently in the PR:

```rust
impl CollatorBorrowed<'_> {
    pub fn write_sort_key<S>(&self, s: &str, sink: &mut S) -> Result<(), core::fmt::Error>
    where
        S: CollationKeySink,
}
```

proposal:

```rust
impl CollatorBorrowed<'a> {
 pub fn sort_key<'b>(&self, s: &'b str) -> impl Iterator<Item = u8> + 'a + 'b
}
```


- @Manishearth I'm not sure how easy it would be to transform this function into a state machine. Looking at the impl, it seems nontrivial. It seems like a whole lot of work.
- @robertbastian I think there are only 3 or 4 places where it writes the sink.
- @Manishearth It's in a loop behind like 5 if blocks.
- @sffc It seems harmless to have `write_sort_key`, and we could add `sort_key` in the future
- @Manishearth The unicode segmentation crate has segmentation as a state machine, and it is a pain to maintain.

Moving to the second question about the shape of the trait:

- @Manishearth If the trait lives in its own crate, it seems fine
- @robertbastian what do we implement this on? if we only do `Vec` and `ShortVec`, this is actually infallible. Do we actually need `&mut [u8]` or can we address this with the iterator API later?
- @echeran I'm thinking at a high level about the iterator versus the sink. A sort key is, each character has primary, secondary, and tertiary weights. It takes the primaries and puts them in a row, secondary and puts them in a row, and so forth. An iterator seems like you could stop reading when you no longer need the elements, but you still need to visit all the code points.
- @robertbastian Yeah, it would require a custom sink impl
- @sffc That's how Writeable comparision works
- @Manishearth We have explicit APIs for doing this type of comparision. You shouldn't use this API to build a comparison.
- @echeran What's the use case for comparing a string to a sort key?
- @sffc That seems like the _main_ use case actually. You have a database and you need to see where the string fits in.
- @echeran Why not just get the sort key of the new string?
- @sffc So you can exit early.
- @sffc Only the Iterator API lets you exit early.
- @robertbastian No, with the sink API, you can return an error to exit early. Which is motivation to have a custom error.
- @manishearth Whenever I have some sort of pluggable callbacky API I'm always frustrated when I can't propagate my errors through it.
    - If you're doing anything more than "write to container" (like comparing against a string) you will need custom errors
    - Even if you're doing "write to container" sometimes the container is a JavaScript string and you need to propagate custom errors.
    - Or even just a String, where you want to catch utf8 errors.

- @sffc Is there a reason to not prefer `CollationKeySink`?
- @robertbastian If there's some other API in `icu_collator` in the future that needs it
- @sffc How about `CollationByteSink`?
- @robertbastian What should we implement this on? Just Vec?
- @sffc Why not mut slice?
- @robertbastian Because then we'd need to return a `usize` so you know how many bytes were written
- @sffc It seems like it's good hygiene to return the `usize` anyway
- @robertbastian Do you like the proposal 2 above?
- @sffc Yeah, `S::Success` is cool. I think it could even be used to return an `Ordering`.
- @Manishearth The `type Success` needs a place to be stored. I think we should make that a follow-up issue.
- @sffc Yeah, OK. I think we should return `usize` as a stopgap.
- @robertbastian I think that's a suboptimal API. Are we stabilizing this or putting it behind an unstable feature?
- @sffc The priority should be landing the PR. It's fine if it's behind an unstable feature.
- @robertbastian `normalize_to` doesn't return a `usize` and doesn't allow writing to anything non-allocating
- @sffc It's an interesting observation. I don't think it's fully transferrable: writing a string to `&mut [u8]` has funny safety invariants.
- @robertbastian Writeable doesn't return the usize.
- @manishearth Since we control the trait (unlike `Write`) we can even provide our own impl on `&mut (&mut [u8], usize)` for convenience.
- @sffc Yes, I like that.
- @robertbastian OK. Do we still need the error?
- @sffc We still want it to short-circuit.
- @Manishearth And if you're writing to something completely different like a String or a JSString.star
- @sffc For the name of the trait, can we keep it `CollationKeySink`, and if we need it for something else in the future, we can rename it `WriteBytes` and add the type alias

</details>

```rust
impl CollatorBorrowed<'a> {
    pub fn write_sort_key_to<S>(&self, s: &str, sink: &mut S) -> Result<(), S::Error>
    where
        S: CollationKeySink,
}

impl CollationKeySink for Vec<u8> { type Error = core::convert::Infallible; }
impl CollationKeySink for ShortVec<u8> { type Error = core::convert::Infallible; }
impl CollationKeySink for VecDecque<u8> { type Error = core::convert::Infallible; }
// maybe later
// impl<'a> CollationKeySink for &'a mut (&'a mut [u8], usize) { type Error = TooSmall; }
```

Conclusion:

- Use the API shape written above, including names of things
- Methods have `_to` suffix

LGTM: @sffc @robertbastian @Manishearth 

@rs-sac Interested in your take on this. We can land the PR with these changes, but if you'd prefer for us to make them we're happy to oblige as well.

rs-sac commented:
The suggestions all seem fine to me.  I have attempted to address them in a new set of patches.

There was a little impedance mismatch when it came to calling the normalizer, which I don't think any of us noticed would happen, because the normalizer has a similar sink which is always fallible, because that code didn't parameterize whether an error is returned via an associated type.  Anyway, have a look at my attempt to handle the problem semi-reasonably.

hsivonen commented:
Sorry about my unavailability for yesterday's meeting.

The reason why the normalizer uses a sink instead of returning a collectable iterator is being able to pass through already-normalized slices. (`icu_normalizer` writing to a `String` is faster than `collect`ing a `unicode-normalization`-returned iterator into a `String`.)

When the input is `&str` or `&[u8]` and `Strength` is `Identical` this passthrough from the normalizer is relevant in collation key generation operation, too.

And, as noted in the meeting (if I'm reading right), since the bytes from the lower strength levels between `Primary` and `Identical` are first buffered, it makes sense to copy them to the caller-visible sink via slices.

robertbastian commented on components/collator/src/comparison.rs:
```diff
@@ -2145,32 +2152,126 @@ impl CollatorBorrowed<'_> {
+    // which is impossible.
+    //
+    // This shape is so it fits neatly into Result::map_err.
+    fn error(err: core::fmt::Error) -> Self::Error;
```
where is this used?

robertbastian reviewed (commented):
Great! Just two small things

echeran commented:
The conclusion of the discussion after I left looks good. I asked the nearest collation expert I know about whether it makes sense to compare a string to a sort key, and here's what I learned from that discussion. I think the ICU4X meeting discussion conclusion aligns with what I learned, so I'm just relaying this information to provide more context for any future discussions going forward:

* The usual way to compare strings in collator is to compare the Collation Elements of the respective strings. You can't compare strings directly. You have to firstly fetch the Collation Elements of the strings.
* Performing comparison follows the collation algorithm: from the Collation Elements, you compare all the primary weights first, then you compare all the secondary weights, then tertiary weights, etc. You exit early when you get to a difference.
* The point of sort keys is to allow for an optimization based on a time-space tradeoff: if you perform comparisons on the same strings enough, then it might eventually become worth it to create a version of the Collation Elements that can be bytewise compared (faster). That new version/representation is the sort key, of course. Of course, that sort key has to be stored somewhere (more space).
* DBs are one obvious place where the comparison of strings may happen over and over again, but that's not necessarily the only place where this optimization tradeoff would make sense.
* There is no early exit when constructing a sort key. Constructing a sort key requires you to visit the whole string because you have to visit all of the primary weights (and serialize them) before visiting all of the secondary weights, and then the tertiary weights, and so on. You have to process the entire string to construct that accurately. When you think about how sort keys get used, which is serializing to a persistent storage for reuse, it makes sense that you need to process the entire string before storing it.
* ICU will compress the bytes that it creates for a sort key. It does this by having knowledge of how the Collation Element weights are constructed so that the compression still preserves the outcome.
* Comparing a string to a sort key doesn't make sense. Instead, you would naturally convert the string into a sort key and then do the easy bytewise comparison on the other sort key. Trying to compare a string to a sort key without converting the string into its own sort key implies that you would be doing duplicate work: you would need to reconstitute Collation Elements from the compressed bytes of the sort key somehow, and compare them to the Collation Elements of the string. But the sort key already is the product of the work to get the Collation Elements and rearrange the constituent weights for comparison purposes. Furthermore, I'm not sure if it's possible that you could reconstitute the Collation Elements from the compressed bytes of the sort key. So in the end, if you "enter sort key space", you stay there and commit to it. Presumably, you would only have a scenario with a sort key in the first place because of a previous decision to commit to the space-time tradeoff optimization in which sort keys are useful, as justified by the use case that would need to preexist in that scenario.

sffc commented on components/collator/src/comparison.rs:
```diff
@@ -1551,4 +1633,916 @@ impl CollatorBorrowed<'_> {
+///
+/// (This crate does not have access to [`std`].)
+pub trait CollationKeySink {
+    /// The type of error the sink may return.
+    type Error;
+
+    /// An intermediate state object used by the sink, which must implement [`Default`].
+    type State;
+
+    /// A result value indicating the final state of the sink (e.g. a number of bytes written).
+    type Output;
```
Observation: I'm fine with the three-associated-types solution, although I think it is achievable by implementing the trait on a type that itself contains a slot for intermediate state, so adding the associated types for State and Output doesn't in itself unlock new functionality.

sffc reviewed (approved):
Thank you for your great work!

Manishearth commented:
Are we ready to merge this? We have two approvals and @hsivonen approved a while ago too. (I'm reluctant to hit the button here since I haven't been actively involved in this change)

@rs-sac thank you so much for bearing with us as we figured this out, and thanks for all of the work!

hsivonen commented:
Merged. Thank you for the code and thank you for your patience with the review comments.

sffc commented:
For posterity, some belated notes from 2025-05-15 ICU4X WG about the Write trait:

- @hsivonen There's a new trait called `Write`. Unfortunately there isn't a trait of exactly the right shape in `core`. So we need to mint a new trait. Should we call it `Write` so that it signals that it is similar to `std::io::Write`, or should we call it something like `CollationKeySink`? And should the methods return a `Result` like the `core`/`std` ones do? Is it useful to allow for fallible sink implementations? In the Normalizer use case, we use regular `core::fmt::Write` which is fallible.
- @sffc I prefer `CollationKeySink`. If it were named `Write`, I feel like it should go into a crate like `core_bytes_write` which I'm not really interested in making. It seems fine for this to be collator-specific.
- @sffc I think it should be fallible like all the other Write traits. Maybe you're serializing into a fixed-length field in a database and you want to know when you run out of room.
- @sffc Maybe use `core::fmt::Error` as the error type instead of creating a new one?
- @hsivonen Yeah that aligns with how we handle strings internally (?)
